#3167 Fixed escaping fields

Patrykb0802 opened the pull request.

Limraj committed:
#3174 Updated REST API with incoming data for XSS protection:
- Updated use of @XssProtect annotation;
- Corrected definition tests in build.gradle for XssValidatorUtilsTestsSuite;

Limraj committed:
#3174 Updated REST API with incoming data for XSS protection:
- Corrected UserProfilesAPI.java

Limraj committed:
#3174 Updated REST API with incoming data for XSS protection:
- Corrected EventDetectorDTO.java;

Limraj committed:
#3174 Updated REST API with incoming data for XSS protection:
- Added support for List and Map for XssProtect annotations;
- Corrected protect in: FolderPointHierarchy, CompoundComponentDTO;

Limraj committed:
#3174 Updated REST API with incoming data for XSS protection:
- Added support array for XssProtect annotation;
- Corrected protect for JsonEventSearch;

Limraj committed:
#3178 Hide user information in DWR:
- Fixed select 'Add user' in MailingList;
- Added: UserIdentifier, AddressEntryJson, EmailRecipientJson, MailingListJson, UserEntryJson;
- Reduced data users in: ViewDwr, WatchListDwr, UsersDwr, ReportsDwr, MailingListsDwr, EventHandlersDwr, DataPointDetailsController;

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.0' into fix/#3178_Hide_user_information_in_DWR

Limraj committed:
#3178 Hide user information in DWR:
- Corrected MailingListsDwr.getMailingList(int id);
- ReportsDwr.init;

Patrykb0802 committed:
#3167 Fixed escaping fields

Patrykb0802 commented on WebContent/WEB-INF/jsp/pointEdit/pointProperties.jsp:
```diff
@@ -62,7 +62,7 @@
     <spring:bind path="form.description">
```
spring:bind already returns an HTML-escaped status.value. Wrapping it with <c:out> causes double escaping, which results in showing &l t;script&g t; instead of <script> in the input field. That’s why we keep only ${status.value} inside spring:bind, without <c:out>.

Limraj replied:
For all tags: spring:bind then set false for 'htmlEscape' attribute;

https://docs.spring.io/spring-framework/docs/1.2.6/taglib/tag/BindTag.html

For this you have to use <c:out> everywhere, so that there is one approach to escaping such value.

Patrykb0802 committed:
#3167 Fixed escaping fields

Limraj commented on WebContent/WEB-INF/jsp/compoundEvents.jsp:
```diff
@@ -38,7 +38,7 @@
             for (i=0; i<data.dataPoints.length; i++) {
                 dp = data.dataPoints[i];
-                pointNode = dojo.widget.createWidget("TreeNode", {title: "<img src='images/icon_comp.png'/> "+ dp.name});
+                pointNode = dojo.widget.createWidget("TreeNode", {title: "<img src='images/icon_comp.png'/> "+ escapeHtml(dp.name)});
```
Escaping on the frontend side does nothing, because before the escapeHtml function is executed on the unescapable value, the code can be executed by the browser, so we cannot prevent it this way.

You need to revert the changes in compoundEvents.jsp . Instead, you need to add converters for the EventSourceBean, EventTypeVO and EventHandlerVO types to dwr.xml for String fields, then dwr will return escaped structures. Similarly to the DataPointVO type and the XssDataPointVoConverter converter.

Limraj commented on WebContent/WEB-INF/jsp/eventHandlers.jsp:
```diff
@@ -228,7 +228,7 @@
 
         var node = dojo.widget.createWidget("TreeNode", {
-                title: "<img src='"+ img +"'/> <span id='"+ handler.id +"Msg'>"+ handler.message +"</span>",
+                title: "<img src='"+ img +"'/> <span id='"+ handler.id +"Msg'>"+ escapeHtml(handler.message) +"</span>",
```
You need to revert the changes in eventHandlers.jsp . If a dwr converter is added for the EventHandlerVO type, this change in the jsp will not be necessary.

Limraj commented on WebContent/WEB-INF/jsp/mailingLists.jsp:
```diff
@@ -164,7 +164,7 @@
 
     function updateMailingList(ml) {
-        $set("ml"+ ml.id +"Name", ml.name);
+        $set("ml"+ ml.id +"Name", escapeHtml(ml.name));
```
Similarly to before, you need to add dwr converters for the types: MailingList, AddressEntry, User. (These are classes that have string fields.)

Limraj commented on WebContent/WEB-INF/jsp/pointHierarchySLTS.jsp:
```diff
@@ -817,12 +817,13 @@ var messages = {
  		          $button.spin();
  		          dialog.setClosable(false);
- 		          $.ajax({
+ 		          const rawInput = dialog.getModalBody().find('input').val();
```
I would use const for constants (a variable that does not change while the program is running), let works better in the scope of a function block.

Limraj commented on WebContent/WEB-INF/jsp/pointHierarchySLTS.jsp:
```diff
@@ -817,12 +817,13 @@ var messages = {
 		            type: "POST",
 		        	dataType: "json",
-		        	url:myLocation+"pointHierarchy/new/0/"+dialog.getModalBody().find('input').val(),
+		        	url:myLocation+"pointHierarchy/new/0/"+encodeURIComponent(rawInput),
```
Here you just need to remove the / and \ characters which will be read as another part of the url address, the rest will be automatically escaped.

Limraj commented on WebContent/WEB-INF/jsp/pointHierarchySLTS.jsp:
```diff
@@ -817,12 +817,13 @@ var messages = {
 		        	success: function(msg){
-		        	  var titleNewNode = dialog.getModalBody().find('input').val();
+		        	  var titleNewNode = escapeHtml(rawInput);
 		        	  dialog.getModalBody().html('<div><h3>'+messages.folder+':</h3><ul><li>'+messages.key+':<b>'+msg+'</b></li><li>'+messages.title+':<b>'+titleNewNode+'</b></li></ul></div>');
```
And removed line:
```
dialog.getModalBody().html('<div><h3>'+messages.folder+':</h3><ul><li>'+messages.key+':<b>'+msg+'</b></li><li>'+messages.title+':<b>'+titleNewNode+'</b></li></ul></div>');
```
This dialog box is closed immediately, so the user has no chance to see the contents of this window anyway, and it does not contain any particularly important information, so I suggest deleting it.

So I propose the following change:
```
 		          $button.spin();
 		          dialog.setClosable(false);
 		          let inputs = dialog.getModalBody().find('input');
 		          let titleNewNode;
 		          if(inputs.length == 1) {
 		               let inputNode = inputs[0];
                       titleNewNode = inputNode.value ? inputNode.value.replaceAll('\\','').replaceAll('\/','') : '';
 		          }
 		          
 		          if(!titleNewNode) {
 		            dialog.getModalBody().html('<div><h3>'+messages.folderNotAdd+'</h3><p>'+ messages.errorThrown +':'+errorThrown+'</p></div>');
 		            dialog.setClosable(true);
                    dialog.getButton('btn-Close').enable();
 		            return;
 		          }
 		          
 		          $.ajax({
		            type: "POST",
		        	dataType: "json",
		        	url:myLocation+"pointHierarchy/new/0/"+titleNewNode,
		        	success: function(msg){
		        	  $button.hide();
		 		      $button.stopSpin();
```

Limraj commented on WebContent/WEB-INF/jsp/publisherList.jsp:
```diff
@@ -34,7 +34,7 @@
         dwr.util.addRows("publisherList", publishers,
             [
-                function(p) { return "<b>"+ p.name +"</b>"; },
+                function(p) { return "<b>"+ escapeHtml(p.name) +"</b>"; },
```
Similarly to the previous step, create a dwr converter for the type: PublisherVO. Change in jsp file to be rolled back.

Limraj commented on WebContent/resources/common.js:
```diff
@@ -1320,7 +1320,6 @@ function unescapeHtml(value) {
 function escapeHtml(value) {
    let div = document.createElement("div");
    div.textContent = value;
-   div.innerText = value;
```
Why was this line removed?

Limraj commented on WebContent/WEB-INF/jsp/watchList.jsp:
```diff
@@ -658,10 +658,10 @@
               <td class="smallTitle"><spring:message code="watchlist.watchlist"/> <tag:help id="watchList"/></td>
               <td align="right">
-                <sst:select id="watchListSelect" value="${selectedWatchList}" onchange="watchListChanged()"
+                <sst:select id="watchListSelect" value="<c:out value='${selectedWatchList}'/>" onchange="watchListChanged()"
```
Super. That's the point, where you can use <c:out>

Limraj commented on WebContent/WEB-INF/jsp/watchList.jsp:
```diff
@@ -127,7 +127,7 @@
       function addFolder(folder, parent) {
           var folderNode = dojo.widget.createWidget("TreeNode", {
-                  title: "<img src='images/folder_brick.png'/> "+ folder.name,
+                  title: "<img src='images/folder_brick.png'/> "+ escapeHtml(folder.name),
```
Here, similarly as before, to withdraw and create a dwr converter. Roll back changes to the jsp file.

Limraj commented on WebContent/WEB-INF/jsp/usersProfiles.jsp:
```diff
@@ -288,7 +288,7 @@
     
     function updateUserProfile(userProfile) {
-        $("u"+ userProfile.id +"UserProfileName").innerHTML = userProfile.name;
+        $("u"+ userProfile.id +"UserProfileName").textContent = userProfile.name;
```
Here, similarly as before, to withdraw and create a dwr converter. Roll back changes to the jsp file.

Limraj commented on WebContent/WEB-INF/jsp/scheduledEvents.jsp:
```diff
@@ -150,7 +150,7 @@
     
     function updateScheduledEvent(se) {
-        $("se"+ se.id +"Name").innerHTML = se.description;
+        $("se"+ se.id +"Name").textContent = se.description;
```
Here, similarly as before, to withdraw and create a dwr converter. Roll back changes to the jsp file.

Limraj reviewed (commented):
The task involves escaping all possible values ​​in JSP pages. This includes, for example, all forms from the Data Source. This task is quite extensive, so I recommend creating a separate issue for individual pages that were not affected by the changes in this pullrequest. The same applies to publishers, etc.

**It's worth trying to create one converter for the String type, instead of creating separate converters for each complex type, the application will crash, but maybe this is a good direction to improve individual pages later.**

Limraj commented on WebContent/resources/emailRecipients.js:
```diff
@@ -180,7 +180,7 @@ mango.erecip.EmailRecipients = function(prefix, testEmailMessage, mailingLists,
                     list[list.length] = {
                             recipientType: 3, // EmailRecipient.TYPE_ADDRESS
-                            referenceAddress: $(this.prefix + id +"Description").innerHTML};
+                            referenceAddress: $(this.prefix + id +"Description").textContent};
```
Here, similarly as before, to withdraw and create a dwr converter. Roll back changes to the jsp file.

Limraj commented on WebContent/resources/emailRecipients.js:
```diff
@@ -215,7 +215,7 @@ mango.erecip.EmailRecipients = function(prefix, testEmailMessage, mailingLists,
         createFromTemplate(this.prefix +"_TEMPLATE_", id, this.prefix +"List");
         $(this.prefix + id +"Img").src = imgName;
-        $(this.prefix + id +"Description").innerHTML = description;
+        $(this.prefix + id +"Description").textContent = description;
```
Here, similarly as before, to withdraw and create a dwr converter. Roll back changes to the jsp file.

Patrykb0802 committed:
PR #3188: revert client-side escapes; rely on DWR converters (compoundEvents.jsp, eventHandlers.jsp, mailingLists.jsp, emailRecipients.js, common.js)

Patrykb0802 committed:
#3167 Fixed escaping fields

Patrykb0802 committed:
#3167 Fixed escaping fields

Patrykb0802 committed:
#3167 Fixed escaping fields

Patrykb0802 commented:
### PR **#3167 — Fixed Escaping Fields (XSS hardening)**

---

> **TL;DR**
>
> - Eliminated **double-escaping** and **HTML injection** in views.  
> - Added a universal **DWR converter** that **escapes `String` properties by default**, and updated the front-end to **render all user content as text** (never via `innerHTML`).  
> - Result: select lists show readable labels (e.g., ``<script>…</script>`` as **plain text**) and **nothing executes**.

---

### **Key Changes**

#### **Backend**
- **New DWR converter:** `SelectiveXssBeanConverter` *(id: `xssBean`)*
  - Escapes all bean `String` properties (**`& < > " '`**).
  - Handles `LocalizableMessage` → **localize, then escape**.
  - Serializes to `Map<String,Object>`, avoids cycles, delegates to `ConverterManager`.
  - **Opt-out** via annotation **`@AllowHtml`** for explicitly safe HTML fields *(none enabled by default)*.
- **Lightweight HTML escaper:** `XssSanitizer.escape(...)` — predictable escaping of the 5 critical chars.
- **Inbound hardening:** `SanitizingStringEditor` for safe binding of request `String`s.
- **`dwr.xml` registration**
  ```xml
  <converter id="xssBean" class="...SelectiveXssBeanConverter"/>
  ```
  Applied to key VO/beans (e.g., `DataPointVO`, `MailingList`, `UsersProfileVO`, `CompoundEventDetectorVO`, `EventTypeVO`, `EventHandlerVO`, `ReportInstance`, `ReportVO`).  
  Disabled legacy specialized converters **`xssDataPoint*`**.
- **Extra server-side safety:** `PointLinksDwr` escapes `extendedName` when building source/target point lists.

#### **Frontend (JSP/JS)**
- **No `innerHTML` for user content.** Use only:
  - `textContent` / `document.createTextNode(...)`
  - For `<select>` options: **`new Option(label, value)`** (creates a text node)
- **Fix for double-escaped labels in selects (Chosen UI)**
  ```js
  // settingsEditor.jsp
  const label = unescapeHtml(p.name);   // decode once
  jQuery("#settingsPointList").append(new Option(label, p.id)); // text node, not HTML
  ```
- **Replaced `innerHTML` → `textContent`** in sensitive lists/tables:
  `publisherList.jsp`, `scheduledEvents.jsp`, `maintenanceEvents.jsp`, etc.
- **Form views render clean text without double escaping**
  - `pointEdit/pointProperties.jsp`: use `<spring:bind htmlEscape="false">` + `<c:out>` in attributes.
  - Similar adjustments in `mailingLists.jsp`, `usersProfiles.jsp`, `compoundEvents.jsp`.
- **Consistency pass:** `pointHierarchySLTS.jsp` and other hotspots use `escapeHtml`/text-only rendering.

---

### **Before / After**

**Before (unsafe / double-escaped or HTML-interpreting)**
```js
// risk: interprets HTML or shows &lt;...&gt;
el.innerHTML = ri.name;
$("#settingsPointList").append("<option value='" + p.id + "'>" + p.name + "</option>");
```

**After (safe / text-only)**
```js
// safe: text node, no HTML interpretation
el.textContent = ri.name ?? "";
$("#settingsPointList").append(new Option(unescapeHtml(p.name), p.id));
```

**Optional, when a DOM node is required instead of a string**
```js
function renderName(ri) {
  const span = document.createElement("span");
  span.textContent = ri?.name ?? "";
  return span;
}
```

---

### **Why This Works (Security Model)**

1. **Server defaults to escaping** via `xssBean` → browser receives safe entities.  
2. **Client never interprets user HTML:**
   - Inserted as **text nodes only** (`textContent`, `new Option(...)`).
   - `unescapeHtml()` is used **only** to convert entities back to a string **before inserting as text** *(never via `innerHTML`)*.
3. **Controlled exceptions:** if raw HTML is ever needed, a field can be marked with **`@AllowHtml`** *(requires explicit review)*.

---

### **User-Visible Result**

- Names like ``<script>alert('XSS')</script>``:
  - **Render correctly** in selects/lists (human-readable text),
  - **Never execute**, because we do not use `innerHTML`,
  - **No more** `&lt;...&gt;` artifacts (double-escaping removed).

---

### **Testing / Manual Verification**

- Created points/objects with HTML-looking names (``<script>…</script>``, ``<img onerror=…>``).  
- Verified all updated views: labels are **readable**, no `&lt;…&gt;`, **no script execution**.  
- Browser console: no inline event execution, no suspicious warnings.

**Checklist**
- [x] Selects (Chosen) show clean labels  
- [x] Tables/lists render text only  
- [x] Forms bind/display without double-escaping  
- [x] No `innerHTML` for user strings in modified files

---

### **Risks & Follow-ups**

- Audit remaining JSPs for lingering **`innerHTML`**.  
- Only add **`@AllowHtml`** after code review for justified cases.  
- Add unit tests for **`XssSanitizer`** and an integration test covering **`SelectiveXssBeanConverter`**.

---

### **Summary**

This PR establishes a clear end-to-end XSS strategy: **escape by default on the server + render as text on the client**.  
It removes double-escaping artifacts and, more importantly, **closes a class of XSS vectors** caused by accidental `innerHTML` usage and inconsistent escaping.

Limraj reviewed (commented):
1. It is necessary to check whether the use of reflection negatively impacts the performance of this converter. This means that this should be compared to the traditional approach, before changes, for objects converted without using explicit reflection, with a dedicated converter, and with a general converter;

2. It is necessary to maintain a consistent name for the converter, according to the names of previous converters;

3. It is necessary to use the escapeHtml method that was previously used, instead of using different implementations of this method in each class;

4. This is primarily to maintain system stability, and secondarily to protect against XSS, therefore:
a) Recursion cannot be used during this conversion; we check objects to a single level of nesting for the occurrence of the following fields: `String, List<String>, Map<if String, if String> , String[]`. For complex types, a separate conversion should be defined in the dwr.xml file, which gives us control over this process;
b) We use the specific types used in our VO objects, meaning, for example, we check if a field is a String instead of a CharSequence;

5. If we want to use reflection, we focus on annotating fields for VO classes. These are classes whose fields should not change (name changes or field deletion); adding new fields is permitted. Furthermore, not all fields have a setter, they are a combination of several fields;

6. The report name should not be treated as HTML, but I'm open to arguments as to why you allow HTML for the report name.

7. In the converter, you cannot return other types of collections and maps other than the source object.

Overall, this could be a good direction, significantly simplifying the configuration of new classes without the rigid assumption of the existence of specific fields. However, the things I mentioned should be implemented.

What it might look like:
edit 17.10.2025:
[#3167_Fixed_escaping_fields.patch](https://github.com/user-attachments/files/22972257/3167_Fixed_escaping_fields.patch)


The reports page still requires some tweaking. The AllowHtml/NoEscape annotation shouldn't be placed on the name field in ReportVO; this needs to be fixed on the frontend. Next, it would be worth replacing the bean converters with xssBean for the remaining VO types and adjusting the frontend, or where an XSS issue is noticed. Currently, adding a new class simply requires defining a new conversion in dwr.xml. No other backend work should be necessary, but the frontend would need to be adjusted to the escaped value to display it correctly on the page.

Limraj committed:
Merge pull request #3051 from SCADA-LTS/release/2.8.0

Release/2.8.0

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.1' into feature/#3174_Updated_REST_API_with_incoming_data_for_XSS_protection

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.1' into fix/#3178_Hide_user_information_in_DWR

Limraj committed:
#3167 Fixed escaping fields:
- Created universal converter XssBeanConverter;
- Change annotation name from AllowHtml to NoEscape;
- Corrected escaped pointlinks;

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.1' into fix/#3167_Fixed_escaping_fields

Limraj committed:
#3167 Fixed escaping fields:
- Created class XssBeanConverterUtils;
- Added test: XssBeanConverterUtilsTest and class: TestGetExtendedName for test;
- Moved NoEscape and XssBeanConverter to org.scada_lts.web.dwr.security;
- Added converted RecipientListEntryBean and IntValuePair in dwr.xml;

Limraj committed:
Merge pull request #3202 from SCADA-LTS/fix/#3178_Hide_user_information_in_DWR

Fix/#3178 hide user information in dwr

Limraj committed:
Merge branch 'release/2.8.1' into feature/#3174_Updated_REST_API_with_incoming_data_for_XSS_protection

Limraj committed:
Merge pull request #3201 from SCADA-LTS/feature/#3174_Updated_REST_API_with_incoming_data_for_XSS_protection

Feature/#3174 updated rest api with incoming data for xss protection

Limraj committed:
Merge remote-tracking branch 'origin/release/2.8.1' into fix/#3167_Fixed_escaping_fields
